In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pickle

## Processed Data

Exploring the structure of the processed DE-HNN dataset.

In [2]:
idx = 0
with open(f"../data/DEHNN_dataset/{idx}.degree.pkl", "rb") as f:
    dict = pickle.load(f)

In [6]:
dict.keys()

dict_keys(['cell_degrees', 'net_degrees'])

In [19]:
pd.Series(dict['net_degrees']).value_counts().sort_index()

0            857
1          15850
2         525902
3         130515
4          56196
           ...  
2050           2
2139           1
2141           1
4087           1
140605         1
Length: 304, dtype: int64

In [24]:
pd.Series(dict['cell_degrees']).value_counts().sort_index()

0          257
2        99955
3       324927
4       144497
5       217291
6         3999
7         6837
14           1
15           1
30           1
32           6
33           1
50           6
52           1
72           5
75           1
76           3
77           1
78           1
82           2
87           4
91           1
108         24
112          2
113          3
114          1
117         24
118         14
130          1
141          2
143         24
165          2
169          6
177         12
178          1
181          2
310          2
380          6
1243        14
dtype: int64

In [27]:
with open(f"2023-03-06_data/{idx}.bipartite.pkl", "rb") as f:
    dict = pickle.load(f)

In [28]:
dict

{'instance_idx': array([478481, 478482, 478483, ..., 797724, 738342, 797725]),
 'net_idx': array([     0,      0,      0, ..., 821521, 821522, 821522]),
 'edge_attr': array([3, 3, 3, ..., 2, 2, 2]),
 'edge_dir': array([0, 0, 0, ..., 1, 0, 1]),
 'sample_name': '/data/son/Research/chips/congestion_prediction/data/RosettaStone-GraphData-2023-03-06/superblue19/6/',
 'folder': '/data/son/Research/chips/congestion_prediction/data/RosettaStone-GraphData-2023-03-06/superblue1/1/',
 'design': 'superblue1',
 'edge_index': array([[     0,      0,      0, ..., 797724, 821522, 797725],
        [478481, 478482, 478483, ..., 821521, 738342, 821522]])}

In [29]:
with open(f"2023-03-06_data/{idx}.eigen.10.pkl", "rb") as f:
    dict = pickle.load(f)

In [32]:
dict['evals']

array([1.9982028, 1.9984511, 1.9999945, 1.9999961, 1.9999961, 1.999997 ,
       1.999999 , 1.9999994, 2.0000005, 2.0000098], dtype=float32)

In [33]:
dict['evects']

array([[ 1.73489370e-05,  1.81367905e-05, -4.42451652e-04, ...,
        -1.69613311e-04, -1.03775405e-04,  3.63318453e-04],
       [ 2.06509212e-05,  2.13837811e-05, -4.94658831e-04, ...,
        -1.89635786e-04, -1.15998235e-04,  4.06221749e-04],
       [ 2.06760851e-05,  2.14082029e-05, -4.94659995e-04, ...,
        -1.89635917e-04, -1.15999334e-04,  4.06223349e-04],
       ...,
       [-1.21696266e-05, -1.34866805e-05,  3.12872638e-04, ...,
         1.19940436e-04,  7.35051581e-05, -2.57036416e-04],
       [-1.18891930e-05, -1.27987933e-05,  3.12860706e-04, ...,
         1.19941426e-04,  7.34357745e-05, -2.56962288e-04],
       [-1.21403045e-05, -1.31762026e-05,  3.12862045e-04, ...,
         1.19949342e-04,  7.34591595e-05, -2.56991305e-04]], dtype=float32)

In [34]:
with open(f"2023-03-06_data/{idx}.global_information.pkl", "rb") as f:
    dict = pickle.load(f)

In [35]:
dict

{'design': 'superblue1',
 'variant': 1,
 'core_utilization': 0.5,
 'max_routing_layer': 'met5',
 'clk_per': 50,
 'clk_uncertainty': 0.2,
 'flow_stage': 'place_opt',
 'hstrap_layer': 'met5',
 'hstrap_width': 1.6,
 'hstrap_pitch': 27.2,
 'vstrap_layer': 'met4',
 'vstrap_width': 1.5,
 'vstrap_pitch': 27.3}

In [36]:
with open(f"2023-03-06_data/{idx}.metis_part_dict.pkl", "rb") as f:
    dict = pickle.load(f)

In [42]:
list(dict.keys())[:10]

[115, 181, 183, 184, 185, 186, 187, 247, 248, 250]

In [43]:
with open(f"2023-03-06_data/{idx}.net_demand_capacity.pkl", "rb") as f:
    dict = pickle.load(f)

In [48]:
pd.Series(dict['demand']).value_counts() #guessing power demand?

22.000000    8465
21.000000    8238
23.000000    8180
24.000000    7839
20.000000    7553
             ... 
23.381044       1
23.357000       1
22.473118       1
6.394209        1
51.155556       1
Length: 178589, dtype: int64

In [54]:
np.mean(dict['demand']), np.std(dict['demand'])

(21.723934436378457, 7.046322395240629)

In [55]:
with open(f"2023-03-06_data/{idx}.net_features.pkl", "rb") as f:
    dict = pickle.load(f)

In [59]:
len(dict['instance_features'])

821523

In [61]:
len(dict['instance_features'][0])

51

In [63]:
dict['instance_features'][0] #???

array([ 9.8392700e+05,  2.6645090e+02,  0.0000000e+00,  4.1767200e+05,
        1.4504403e+05,  3.1538994e+05,  1.4943203e+05,  3.9212781e+05,
        5.4454250e+05,  1.1225168e+05,  2.7461322e+05,  2.0919844e+05,
        1.7670420e+04,  5.0648119e+05,  8.4914600e+05,  4.0419009e+05,
        1.3343519e+06,  1.4035854e+06,  2.6564066e+05,  1.0652486e+06,
        6.0974419e+05,  4.1504445e+04,  1.2290500e+05,  2.6057900e+05,
        4.7308500e+05,  7.7726800e+05,  1.1596730e+06,  1.8579820e+06,
        4.2378000e+05,  7.0933500e+05,  1.1935630e+06,  1.8414130e+06,
        2.7894380e+06,  4.1105650e+06,  1.9671024e+00,  2.0726898e+00,
       -5.3799072e+01, -6.0643272e+01,  1.4420410e+00, -4.4606724e+00,
       -2.9280889e+00, -2.0650249e+01, -1.2616593e+01,  4.4177708e+01,
        1.7341040e+01,  4.1642575e+00,  5.9796691e-01,  7.7328318e-01,
        3.5756442e-01,  5.6296331e-01,  8.2648188e-01], dtype=float32)

In [64]:
with open(f"2023-03-06_data/{idx}.node_features.pkl", "rb") as f:
    dict = pickle.load(f)

In [65]:
dict

{'num_instances': 797938,
 'num_nets': 821523,
 'x_min': 1920,
 'x_max': 7278720,
 'y_min': 3330,
 'y_max': 7276050,
 'min_cell_width': 1920,
 'max_cell_width': 19211520,
 'min_cell_height': 13320,
 'max_cell_height': 20601600,
 'instance_features': array([[7.77638522e-01, 2.67399267e-01, 1.32000000e+02, 3.99800100e-04,
         0.00000000e+00, 6.00000000e+00],
        [6.15699208e-01, 4.91758242e-01, 3.59000000e+02, 7.99600200e-04,
         0.00000000e+00, 6.00000000e+00],
        [5.91820580e-01, 4.78021978e-01, 1.76000000e+02, 4.99750125e-04,
         0.00000000e+00, 0.00000000e+00],
        ...,
        [7.77506596e-01, 1.34615385e-01, 6.00000000e+00, 1.59920040e-03,
         0.00000000e+00, 6.00000000e+00],
        [7.75593668e-01, 9.34065934e-02, 6.00000000e+00, 1.59920040e-03,
         0.00000000e+00, 0.00000000e+00],
        [7.76781003e-01, 5.86080586e-02, 6.00000000e+00, 1.59920040e-03,
         0.00000000e+00, 0.00000000e+00]]),
 'sample_name': '/data/zluo/new_data/superblue

In [69]:
len(dict['instance_features'][0])

6

In [70]:
dict['instance_features'][0] #???

array([7.77638522e-01, 2.67399267e-01, 1.32000000e+02, 3.99800100e-04,
       0.00000000e+00, 6.00000000e+00])

In general, a bit confused about how the data is layout/partitioned. Is one index a netlist? Whole chip? Additionally, a lot of feature vectors that are not specified exactly what they are in the data writeup.


In [72]:
with open(f"2023-03-06_data/node_neighbors/{idx}.node_neighbor_features.pkl", "rb") as f:
    dict = pickle.load(f)

In [74]:
dict.keys()

dict_keys(['pd', 'neighbor'])

In [75]:
dict['pd']

array([[0.57659997, 3.78525   , 2.84100841, ..., 0.        , 0.        ,
        0.        ],
       [1.1236587 , 2.59016081, 0.48024168, ..., 0.        , 0.        ,
        0.        ],
       [1.1236587 , 2.59016081, 0.48024168, ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.95083978, 0.6301005 ,
        0.03481547],
       [0.        , 0.        , 0.        , ..., 1.14630555, 1.43355668,
        0.14265637],
       [0.        , 0.        , 0.        , ..., 1.44839475, 1.73426691,
        0.1504663 ]])

In [81]:
len(dict['pd'])

797938

In [86]:
len(dict['neighbor'][0]) # Again confused about what these vectors mean, and why do each of pd's and neighbor's elements differ in length

# ok pd is persistant diagram so makes sense for length to be diff

12

In [231]:
idx = 0
info = 'bipartite'
def open_data(idx, info):
    with open(f"2023-03-06_data/{idx}.{info}.pkl", "rb") as f:
        dict = pickle.load(f)
    return dict

### Bipartite

A bipartite representation of a chip, with wires/nets being represented as nodes on top of cells/instances and the pins/terminals being edges.

In [232]:
data = open_data(0, 'bipartite')

In [236]:
data

{'instance_idx': array([478481, 478482, 478483, ..., 797724, 738342, 797725]),
 'net_idx': array([     0,      0,      0, ..., 821521, 821522, 821522]),
 'edge_attr': array([3, 3, 3, ..., 2, 2, 2]),
 'edge_dir': array([0, 0, 0, ..., 1, 0, 1]),
 'sample_name': '/data/son/Research/chips/congestion_prediction/data/RosettaStone-GraphData-2023-03-06/superblue19/6/',
 'folder': '/data/son/Research/chips/congestion_prediction/data/RosettaStone-GraphData-2023-03-06/superblue1/1/',
 'design': 'superblue1',
 'edge_index': array([[     0,      0,      0, ..., 797724, 821522, 797725],
        [478481, 478482, 478483, ..., 821521, 738342, 821522]])}

In [235]:
data.keys()

dict_keys(['instance_idx', 'net_idx', 'edge_attr', 'edge_dir', 'sample_name', 'folder', 'design', 'edge_index'])

In [241]:
len(data['instance_idx']), len(data['net_idx'])

(2950019, 2950019)

'instance_idx', 'net_idx', 'edge_attr', 'edge_dir', and both elements in 'edge_index' have same length (295019)

In [256]:
len(set(data['instance_idx']))

797681

In [263]:
pd.Series(data['instance_idx']).value_counts().iloc[:20]

732495    1243
732475    1243
732516    1243
732509    1243
732484    1243
732505    1243
732479    1243
732514    1243
732490    1243
732507    1243
732481    1243
732519    1243
732497    1243
732503    1243
732458     380
732448     380
732442     380
732464     380
732453     380
732468     380
dtype: int64

In [289]:
pd.Series(pd.Series(data['instance_idx']).value_counts().values).value_counts()

3       324975
5       216932
4       144290
2        99906
7         7016
6         4387
118         38
109         24
144         24
1243        14
178         12
169          6
380          6
32           6
50           6
72           5
87           4
113          3
76           3
142          2
310          2
181          2
82           2
165          2
112          2
75           1
53           1
77           1
33           1
78           1
30           1
15           1
14           1
91           1
179          1
114          1
130          1
dtype: int64

Many duplicate indices within 'instance_idx', not exactly sure where these indices coming from.

Not from cell library data as only 80 different cells within that.

Most indices appear a few times (2-7), though there are certain with specific high frequency (ex. 118, 169, 380, 1243).

No index appears only once.

In [310]:
pd.Series((data['edge_attr'])).value_counts()

1       797681
2       797681
3       697775
4       372800
5       228510
         ...  
671         14
672         14
673         14
674         14
1243        14
Length: 1243, dtype: int64

797681 matches up with the unique number of instance indices.

Makes sense as each instance has at least one pin and this pin must be used.

Tells us each instance has at least 2 pins being used.

In [311]:
len(data['edge_attr'])

2950019

I am assuming edge_attr is the pin/terminal index that is being used by the net/wire to connect to the cell.

In [308]:
data['edge_dir']

array([0, 0, 0, ..., 1, 0, 1])

'edge_dir' only consists of 0s and 1s, assuming signifies which direction the edge is pointing (not exactly sure which direction is for 0 and which for 1).

In [254]:
len(data['edge_index'][1])

2950019

In [291]:
data.keys()

dict_keys(['instance_idx', 'net_idx', 'edge_attr', 'edge_dir', 'sample_name', 'folder', 'design', 'edge_index'])

In [294]:
def get_elem_data(idx):
    print(data['instance_idx'][idx])
    print(data['net_idx'][idx])
    print(data['edge_attr'][idx])
    print(data['edge_dir'][idx])
    print(data['edge_index'][0][idx])
    print(data['edge_index'][1][idx])

In [306]:
get_elem_data(12030)

492078
0
3
0
0
492078
